### PR TITLE
Pipeline Stage added to reduce older lambda versions

### PR DESCRIPTION
**Pull request by timrickwood**
Currently set to retain 5 most recent lambda versions

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+			--name $(VERSION)-$(BUILD_COMMIT_HASH) \
+			--function-name $(NAME) \
+			--function-version $(VERSION) \
+		"
+
+remove-old-versions-for-task: ## Prune old versions of hk task lambdas - Mandatory; [PROFILE] - Optional [TASK]
+	eval "$$(make aws-assume-role-export-variables)"
+	if [ "$(TASK)" == "all" ]; then
+		for task in $$(echo $(TASKS) | tr "," "\n"); do
+			lambda_name="${SERVICE_PREFIX}-hk-$$task-lambda"
```
Comment (timrickwood) [inline Makefile:200]: REview comment extend to handle cron

File: Makefile [hunk 2]
```diff
@@ -2,6 +2,7 @@
 PROJECT_DIR := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
 include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
 
+LAMBDA_VERSIONS_TO_RETAIN = 5
```
Comment (timrickwood) [inline Makefile:5]: Move to profile

Comment (timrickwood) [reply Makefile:5]: moved to nonprod

File: Makefile [hunk 3]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
 
 # --------------------------------------
 
+lambda-alias: ### Updates new lambda version with alias based on commit hash - Mandatory PROFILE=[profile], TASK=[hk task]
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply Makefile:165]: Removed

File: Makefile [hunk 4]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+			make aws-lambda-create-alias NAME=$$function VERSION=$$version
+		done
+	else
+		function=$(SERVICE_PREFIX)-hk-$(TASK)-lambda
+		versions=$$(make -s aws-lambda-get-latest-version NAME=$$function)
+		version=$$(echo $$versions | make -s docker-run-tools CMD="jq '.Versions[-1].Version'" | tr -d '"')
+		make aws-lambda-create-alias NAME=$$function VERSION=$$version
+	fi
+
+aws-lambda-get-latest-version: ### Fetches the latest function version for a lambda function - Mandatory NAME=[lambda function name]
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply Makefile:181]: Removed

File: Makefile [hunk 5]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+	make -s docker-run-tools ARGS="$$(echo $(AWSCLI) | grep awslocal > /dev/null 2>&1 && echo '--env LOCALSTACK_HOST=$(LOCALSTACK_HOST)' ||:)" CMD=" \
+		$(AWSCLI) lambda list-versions-by-function \
+			--function-name $(NAME) \
+			--output json \
+		"
+
+aws-lambda-create-alias: ### Creates an alias for a lambda version - Mandatory NAME=[lambda function name], VERSION=[lambda version]
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply Makefile:188]: Removed

File: Makefile [hunk 6]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+
+
+aws-lambda-remove-old-versions: ## Remove older versions Mandatory NAME=[lambda function] Optional LAMBDA_VERSIONS_TO_RETAIN (default 5)
+	older_versions_to_remove="$$(make aws-lambda-get-versions-to-remove NAME=$(NAME))"
+	# convert space separated string into array
+	version_array=($$older_versions_to_remove)
+	echo "There are $${#version_array[*]} versions to be removed for $(NAME)"
+	for version in $${older_versions_to_remove}
+		do
+			make aws-lamba-function-delete NAME=$(NAME) VERSION=$$version LAMBDA_VERSIONS_TO_RETAIN=$(LAMBDA_VERSIONS_TO_RETAIN)
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply Makefile:218]: Removed

File: Makefile [hunk 7]
```diff
@@ -161,6 +162,102 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+		done
+
+aws-lamba-function-delete: ## Delete version of lambda function - Mandatory NAME=[lambda function] VERSION=[version number]
+	echo "Removing version $(VERSION) for $(NAME)"
+	$$(make -s docker-run-tools ARGS="$$(echo $(AWSCLI) | grep awslocal > /dev/null 2>&1 && echo '--env LOCALSTACK_HOST=$(LOCALSTACK_HOST)' ||:)" CMD=" \
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply Makefile:223]: Removed

File: Makefile (hunk 1 above)
Comment (timrickwood) [reply Makefile:200]: Done

File: Makefile (hunk 2 above)
Comment (timrickwood) [reply Makefile:5]: and copied to put and live profiles

File: Makefile [hunk 8]
```diff
@@ -161,6 +161,72 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
 
 # --------------------------------------
 
+remove-old-versions-for-task: ## Prune old versions of hk task lambdas - Mandatory; [PROFILE] - Optional [TASK]
+	eval "$$(make aws-assume-role-export-variables)"
+	task_type=$$(make task-type NAME=$$task)
```
Comment (dehe1) [inline Makefile:166]: This needs to bee moved inside the for loop as well

File: Makefile [hunk 9]
```diff
@@ -161,6 +161,72 @@ python-code-coverage-format: ### Test Python code with 'coverage' - mandatory: C
...
+	if [ "$(TASK)" == "all" ]; then
+		for task in $$(echo $(TASKS) | tr "," "\n"); do
+			lambda_name="${SERVICE_PREFIX}-$$task_type-$$task-lambda"
+			echo "Checking for older versions of lambda function $$lambda_name"
+			make aws-lambda-remove-old-versions NAME=$$lambda_name
+			done
+	else
+			lambda_name="${SERVICE_PREFIX}-hk-$(TASK)-lambda"
```
Comment (dehe1) [inline Makefile:174]: Also updated and remove reference to hk here

File: Makefile (hunk 8 above)
Comment (timrickwood) [reply Makefile:166]: more haste less speed

File: Makefile (hunk 9 above)
Comment (timrickwood) [reply Makefile:174]: more haste less speed

**Review by dehe1 — APPROVED** [review]

**Review by tamaragoldschmidt — APPROVED** [review]

**Pull request merged**